### ISSUE TITLE
Bug: false positive from no-useless-assignment in Vue SFC

### ISSUE THREAD

**Issue by crystalfp** [eslint/eslint#20486]
### Environment

Node version: v25.5.0
npm version: v11.9.0
Local ESLint version: v10.0.0 (Currently used)
Global ESLint version: v10.0.0
Operating System: win32 10.0.26200

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```js
import {defineConfig} from "eslint/config";
import tsParser from "@typescript-eslint/parser";
import vueParser from "vue-eslint-parser";
import vuePlugin from "eslint-plugin-vue";
import promisePlugin from "eslint-plugin-promise";
import importPlugin from "eslint-plugin-import";
import typescriptPlugin from "@typescript-eslint/eslint-plugin";
import unicornPlugin from "eslint-plugin-unicorn";
import securityPlugin from "eslint-plugin-security";
import sonarjsPlugin from "eslint-plugin-sonarjs";
import regexpPlugin from "eslint-plugin-regexp";
import jsPlugin from "@eslint/js"
import stylistic from "@stylistic/eslint-plugin";
import depend from "eslint-plugin-depend";

export default defineConfig([
    stylistic.configs.customize({
        flat: true, // required for flat config
        indent: 4,
        quotes: "double",
        semi: true,
        jsx: false,
        blockSpacing: false,
        arrowParens: true,
    }), {
    ignores: [
        "src/vite-env.d.ts",
	"src/assets",
        "src/cpp",
        "src/electron/fingerprint/rollup.config.mjs",
        "src/electron/fingerprint/KDtree.js",
    ]}, {
    files: [
        "src/**/*.ts",
        "src/**/*.vue",
        "eslint.config.mjs",
        "vite.config.mts"
    ],
    languageOptions: {
        parser: vueParser,
        ecmaVersion: "latest",
        globals: {
            document: "readonly",
            window: "readonly",
            NodeJS: "readonly",
            test: "readonly",
            afterAll: "readonly",
            beforeAll: "readonly",
            describe: "readonly",
            process: "readonly",
            requestAnimationFrame: "readonly",
            setTimeout: "readonly",
            setInterval: "readonly",
            clearInterval: "readonly",
            clearTimeout: "readonly",
            console: "readonly",
            Buffer: "readonly",
            structuredClone: "readonly",
            addEventListener: "readonly",
        },
        parserOptions: {
            parser: tsParser,
            ecmaFeatures: {impliedStrict: true},
            projectService: {
                defaultProject: "tsconfig.json",
            },
            warnOnUnsupportedTypeScriptVersion: false,
            extraFileExtensions: [".vue"],
            vueFeatures: {filter: false},
        }
    },
    processor: vuePlugin.processors[".vue"],
    linterOptions: {
        reportUnusedDisableDirectives: "error",
        reportUnusedInlineConfigs: "error"
    },
    plugins: {
        vue: vuePlugin,
        promise: promisePlugin,
        import: importPlugin,
        "@typescript-eslint": typescriptPlugin,
        unicorn: unicornPlugin,
        security: securityPlugin,
        sonarjs: sonarjsPlugin,
        regexp: regexpPlugin,
        "@stylistic": stylistic,
        depend: depend,
    },
    settings: {
        "import/parsers": {"@typescript-eslint/parser": [".ts", ".tsx", ".mts"]},
        "import/extensions": [".js", ".ts", ".vue", ".mts", ".mjs"],
        "import/ignore": ["node_modules"],
        "import/resolver": {
            "typescript": {"alwaysTryTypes": true},
            "alias": {"map": {"@/": "./src/"},
            "extensions": [".vue", ".ts"]}
        }
    },
    rules: {
        ...jsPlugin.configs.recommended.rules,
        ...promisePlugin.configs.recommended.rules,
        ...importPlugin.configs.recommended.rules,
        ...importPlugin.configs.typescript.rules,
        ...unicornPlugin.configs.all.rules,
        ...securityPlugin.configs.recommended.rules,
        ...securityPlugin.configs["recommended-legacy"].rules,
        ...sonarjsPlugin.configs.recommended.rules,
        ...regexpPlugin.configs.recommended.rules,
        ...vuePlugin.configs["flat/essential"].rules,
        ...vuePlugin.configs["flat/strongly-recommended"].rules,
        ...vuePlugin.configs["flat/recommended"].rules,
        ...typescriptPlugin.configs.recommended.rules,
        ...typescriptPlugin.configs["recommended-type-checked"].rules,
        ...typescriptPlugin.configs["stylistic-type-checked"].rules,
        ...typescriptPlugin.configs["strict-type-checked"].rules,
        ...depend.configs["flat/recommended"].rules,

        // > ******************* language rules ***********************
        "@typescript-eslint/consistent-type-assertions": ["warn", {assertionStyle: "as"}],
        "@typescript-eslint/array-type": ["warn", {default: "array", readonly: "array"}],
        "@typescript-eslint/no-unnecessary-condition": "off",
        "@typescript-eslint/restrict-template-expressions": ["warn", {allowNumber: true}],
        "no-dupe-class-members": "error",
        "no-redeclare": ["off", {builtinGlobals: true}],
        "@typescript-eslint/no-redeclare": ["error", {builtinGlobals: true}],
        "no-unused-vars": "off",
        "@typescript-eslint/no-unused-vars": ["error", {
            "vars": "all",
            "args": "all",
            "argsIgnorePattern": "^_",
            "caughtErrors": "all",
            "caughtErrorsIgnorePattern": "^_",
            "destructuredArrayIgnorePattern": "^_",
            "varsIgnorePattern": "^_",
            "ignoreRestSiblings": true
        }],
        "no-empty-function": "warn",
        "no-useless-constructor": "warn",
        "no-loop-func": "warn",
        "no-unused-expressions": "warn",
        "no-shadow": "off",
        "@typescript-eslint/no-shadow": [
            "error", {hoist: "functions-and-types"}
        ],
        "no-restricted-syntax": ["error", "BinaryExpression[operator='in']"],
        "eqeqeq": ["error", "always"],
        "no-unused-private-class-members": "error",
        "prefer-const": "warn",
        "camelcase": ["warn", {properties: "never"}],
        "no-confusing-arrow": ["error", {allowParens: true}],
        "no-else-return": "error",
        "no-array-constructor": "error",
        "require-await": "off",
        "@typescript-eslint/require-await": "error",
        "no-return-await": "off",
        "@typescript-eslint/return-await": "error",
        "no-negated-condition": "warn",
        "no-invalid-this": "off",
        "@typescript-eslint/no-invalid-this": ["error", {capIsConstructor: false}],
        "@typescript-eslint/explicit-function-return-type": ["warn", {allowExpressions: true}],
        "@typescript-eslint/prefer-includes": "warn",
        "@typescript-eslint/prefer-nullish-coalescing": "warn",
        "@typescript-eslint/prefer-optional-chain": "warn",
        "@typescript-eslint/no-confusing-void-expression": ["error", {ignoreArrowShorthand: true}],
        "default-param-last": "warn",
        "@typescript-eslint/no-non-null-assertion": "off",
        "no-loss-of-precision": "error",
        "no-duplicate-imports": ["error", {"allowSeparateTypeImports": true}],
        "@typescript-eslint/await-thenable": "error",
        "@typescript-eslint/no-array-delete": "error",
        "@typescript-eslint/no-base-to-string": "error",
        "@typescript-eslint/no-duplicate-type-constituents": "error",
        "@typescript-eslint/no-floating-promises": "error",
        "@typescript-eslint/no-for-in-array": "error",
        "@typescript-eslint/no-implied-eval": "error",
        "@typescript-eslint/no-meaningless-void-operator": "error",
        "@typescript-eslint/no-misused-promises": "error",
        "@typescript-eslint/no-misused-spread": "error",
        "@typescript-eslint/no-mixed-enums": "error",
        "@typescript-eslint/no-redundant-type-constituents": "error",
        "@typescript-eslint/no-unnecessary-boolean-literal-compare": "error",
        "@typescript-eslint/no-unnecessary-template-expression": "error",
        "@typescript-eslint/no-unnecessary-type-arguments": "error",
        "@typescript-eslint/no-unnecessary-type-assertion": "error",
        "@typescript-eslint/no-unsafe-argument": "error",
        "@typescript-eslint/no-unsafe-assignment": "error",
        "@typescript-eslint/no-unsafe-call": "error",
        "@typescript-eslint/no-unsafe-enum-comparison": "error",
        "@typescript-eslint/no-unsafe-member-access": "error",
        "@typescript-eslint/no-unsafe-return": "error",
        "@typescript-eslint/no-unsafe-type-assertion": "off",
        "@typescript-eslint/no-unsafe-unary-minus": "error",
        "@typescript-eslint/non-nullable-type-assertion-style": "error",
        "@typescript-eslint/only-throw-error": "error",
        "@typescript-eslint/prefer-promise-reject-errors": "error",
        "@typescript-eslint/prefer-reduce-type-parameter": "error",
        "@typescript-eslint/prefer-return-this-type": "error",
        "@typescript-eslint/promise-function-async": "error",
        "@typescript-eslint/related-getter-setter-pairs": "error",
        "@typescript-eslint/require-array-sort-compare": "error",
        "@typescript-eslint/restrict-plus-operands": "error",
        "@typescript-eslint/switch-exhaustiveness-check": "error",
        "@typescript-eslint/unbound-method": "error",
        "@typescript-eslint/use-unknown-in-catch-callback-variable": "off",
        "no-dupe-keys": "error",
        "@typescript-eslint/no-useless-default-assignment": "off",

        // > ******************* stylistic ***********************
        "@stylistic/lines-between-class-members": "warn",
        "@stylistic/quotes": ["warn", "double", {avoidEscape: true}],
        "@stylistic/space-before-function-paren": [
            "error", {anonymous: "never", named: "never", asyncArrow: "always", catch: "never"}
        ],
        "@stylistic/comma-spacing": ["error", {before: false, after: true}],
        "@stylistic/keyword-spacing": ["warn", {before: true, after: true, overrides: {
            if: {after: false},
            switch: {after: false},
            catch: {after: false},
            while: {after: false},
            for: {after: false},
        }}],
        "@stylistic/object-curly-spacing": ["warn", "never"],
        "@stylistic/brace-style": ["warn", "stroustrup", {allowSingleLine: true}],
        "@stylistic/no-multi-spaces": "off",
        "@stylistic/multiline-ternary": "off",
        "@stylistic/operator-linebreak": "off",
        "@stylistic/indent-binary-ops": "off",
        "@stylistic/key-spacing": "off",
        "@stylistic/max-statements-per-line": ["warn", {"max": 2}],
        "@stylistic/comma-dangle": ["off", {arrays: "only-multiline", objects: "only-multiline"}],
        "@stylistic/spaced-comment": ["warn", "always", {markers: [":", "-", "+", "::", "/"]}],
        "@stylistic/space-infix-ops": "off",
        "@stylistic/semi-spacing": "error",
        "@stylistic/no-extra-semi": "error",
        "@stylistic/no-extra-parens": ["warn", "functions"],
        "@stylistic/semi": ["error", "always"],
        "@stylistic/no-multiple-empty-lines": ["warn", {max: 2, maxEOF: 1, maxBOF: 0}],
        "@stylistic/padded-blocks": "off",
        "@stylistic/indent": "off",
        "@stylistic/no-tabs": "off",
        "@stylistic/no-mixed-spaces-and-tabs": "off",
        "@stylistic/member-delimiter-style": "warn",
        "@stylistic/array-bracket-newline": ["warn", "consistent"],
        "@stylistic/array-bracket-spacing": ["warn", "never"],
        "@stylistic/arrow-parens": ["warn", "always"],
        "@stylistic/yield-star-spacing": ["error", {after: true, before: false}],
        "@stylistic/generator-star-spacing": ["error", {after: true, before: false}],
        "@stylistic/eol-last": ["error", "always"],
        "@stylistic/new-parens": "error",
        "@stylistic/quote-props": ["warn", "as-needed"],

        // > ******************* unicorn ***********************
        "unicorn/numeric-separators-style": ["off", {number: {onlyIfContainsSeparator: true, minimumDigits: 3}}],
        "unicorn/prevent-abbreviations": ["warn", {
                replacements: {
                    len: false,
                    params: false,
                    doc: false,
                    pkg: false,
                    ctx: false,
                    i: false,
                    j: false,
                    idx: false,
                    args: false,
                    dir: false,
                    props: false,
                    db: false,
                    obj: false,
                    ext: false,
                    dist: false,
                    dev: false,
                    temp: false,
                    e: false
                },
                checkFilenames: false
            }
        ],
        "unicorn/prefer-top-level-await": "off",
        "unicorn/no-zero-fractions": "off",
        "unicorn/no-for-loop": "off",
        "unicorn/prefer-import-meta-properties": "off",
        "unicorn/filename-case": ["off", {case: "camelCase"}],
        "unicorn/throw-new-error": "off",
        "unicorn/new-for-builtins": "off",
        "unicorn/switch-case-braces": "off",
        "unicorn/prefer-json-parse-buffer": "off",
        "unicorn/prefer-class-fields": "warn",
        // === Enable non-recommended rules that add value ===
        'unicorn/better-regex': 'warn',              // Simplify regexes: /[0-9]/ → /\d/
        'unicorn/custom-error-definition': 'error',  // Correct Error subclassing
        'unicorn/no-unused-properties': 'warn',      // Dead code detection
        'unicorn/consistent-destructuring': 'warn',  // Use destructured vars consistently

        // > ******************* sonarjs ***********************
        "sonarjs/cognitive-complexity": ["off", 40],
        "sonarjs/no-commented-code": "warn",
        "sonarjs/void-use": "off",
        "sonarjs/different-types-comparison": "off",
        "sonarjs/no-unenclosed-multiline-block": "off",
        "sonarjs/fixme-tag": "off",
        "sonarjs/no-labels": "off",
        "sonarjs/todo-tag": "off",
        "sonarjs/pseudo-random": "off",
        "sonarjs/no-selector-parameter": "off",

        // > ******************* other plugins ***********************
        "depend/ban-dependencies": ["warn", {
                                    "presets": ["native", "microutilities", "preferred"]
        }],
        "vue/v-bind-style": ["warn", "shorthand", {"sameNameShorthand": "always"}],
        "vue/attributes-order": "warn",
        "vue/no-undef-components": ["error", {"ignorePatterns": ["^v-", "^router-"]}],
        "vue/prefer-use-template-ref": "error",
        "vue/no-unused-properties": ["error", {
            "groups": ["props", "data", "computed", "methods"]
        }],
        "vue/no-unused-refs": "error",
        "vue/no-unused-emit-declarations": "error",
        "import/no-unresolved": "error",
        "import/default": "off",
        "promise/always-return": "off",
        "promise/catch-or-return": ["warn", {allowFinally: true}],
        "security/detect-non-literal-fs-filename": "off",
        "security/detect-object-injection": "off",
        "@typescript-eslint/no-deprecated": "off",
        "security/detect-unsafe-regex": "off",
        "security/detect-no-csrf-before-method-override": "off",
        "@typescript-eslint/consistent-generic-constructors": "off",
    }
}]);
```
</details>

```js
<script setup lang="ts">
const isDevelopment = import.meta.env.DEV;
</script>

<template>
  <div v-if="isDevelopment">Currently running in the development environment</div>
</template>
```

### What did you expect to happen?

No complain from the `no-useless-assignment` rule as was in 9.x

### What actually happened?

```
error: This assigned value is not used in subsequent statements (no-useless-assignment) at src\components\About.vue:21:7:
  1 |
> 2 | const isDevelopment = import.meta.env.DEV;
     |       ^
  3 |
```

### Link to Minimal Reproducible Example

The code is above

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by crystalfp** [comment]
Sorry, the parser is `vue-eslint-parser`

**Comment by crystalfp** [comment]
https://play.vuejs.org/#eNp9UbFOAzEM/RUrE0jQDjBV10pQOsAACBizVHe+NiXnRIlzFFX37zipWoqEujl+L8/v2Tt15/2oT6gmqop1MJ4hIic/02Q67wLDLmA7QBtcB1oJUytNtaPIYOID9mid75AYpiDECw4JLzVV472YyMiDsfN2ySgvgKoxPfTXpp1q9UdBq9k8hSCV/YaQiAytwBDwGqE5GYTUm+Ao19VYxES1Gp+MUFeKozhszWq0iY4k2i4P1qp2nTcWw4tnIwm0mkBBMra01n09lV7OcHXo12usP//pb+I297R6DRgx9LKXI8bLsELJk+HF+zNupT6CnWuSFfYZ8A2jsyl73NPuEzVi+4RX3D6WA8mWPuJiy0jxECobzcyh8MvV5mei/9q9Gd2Wf5oGNfwAgDe3rw==

**Comment by mdjermanovic** [comment]
Hi @crystalfp, thanks for the issue!

Can you please provide a minimal reproducible example (StackBlitz or GitHub repo) so we can take a look.

**Comment by crystalfp** [comment]
Added in my 2nd comment

**Comment by mdjermanovic** [comment]
> Added in my 2nd comment

I see the code in the Vue Playground, but it doesn't reproduce the reported problem with ESLint. Can you provide a minimal repro where we could run ESLint and debug to find out what is exactly causing this problem?

**Comment by barisusakli** [comment]
I am getting the same error ` 14:6  error  This assigned value is not used in subsequent statements  no-useless-assignment`

I created a minimal repo https://github.com/barisusakli/eslitn10-test.

It complains about the variable `groupTitles` , If I comment out the line `groupTitles = JSON.parse(data.groupTitle);` it passes.
```
function isJSON(str) {
	return !!str;
}

function isGroupTitleValid(data) {
	function checkTitle(title) {
		if (title === 'registered-users') {
			throw new Error('[[error:invalid-group-title]]');
		}
	}
	if (!data.groupTitle) {
		return;
	}
	let groupTitles = [];
	if (isJSON(data.groupTitle)) {
		groupTitles = JSON.parse(data.groupTitle);
		if (!Array.isArray(groupTitles)) {
			throw new Error('[[error:invalid-group-title]]');
		}
		groupTitles.forEach(title => checkTitle(title));
	} else {
		groupTitles = [data.groupTitle];
		checkTitle(data.groupTitle);
	}
	if (groupTitles.length > 1) {
		data.groupTitle = JSON.stringify(groupTitles[0]);
	}
}

isGroupTitleValid({ groupTitle: '["registered-users", "guests"]' });
```

**Comment by mdjermanovic** [comment]
> I am getting the same error ` 14:6 error This assigned value is not used in subsequent statements no-useless-assignment`
> 
> I created a minimal repo https://github.com/barisusakli/eslitn10-test.
> 
> It complains about the variable `groupTitles` , If I comment out the line `groupTitles = JSON.parse(data.groupTitle);` it passes.
> 
> ```
> function isJSON(str) {
> 	return !!str;
> }
> 
> function isGroupTitleValid(data) {
> 	function checkTitle(title) {
> 		if (title === 'registered-users') {
> 			throw new Error('[[error:invalid-group-title]]');
> 		}
> 	}
> 	if (!data.groupTitle) {
> 		return;
> 	}
> 	let groupTitles = [];
> 	if (isJSON(data.groupTitle)) {
> 		groupTitles = JSON.parse(data.groupTitle);
> 		if (!Array.isArray(groupTitles)) {
> 			throw new Error('[[error:invalid-group-title]]');
> 		}
> 		groupTitles.forEach(title => checkTitle(title));
> 	} else {
> 		groupTitles = [data.groupTitle];
> 		checkTitle(data.groupTitle);
> 	}
> 	if (groupTitles.length > 1) {
> 		data.groupTitle = JSON.stringify(groupTitles[0]);
> 	}
> }
> 
> isGroupTitleValid({ groupTitle: '["registered-users", "guests"]' });
> ```

This works as intended because the `[]` value assigned on initialization is never used as the variable is reassigned at the beginning of both `if` and `else` blocks. If you replace `let groupTitles = [];` with `let groupTitles;` it doesn't change the behavior.

I also checked this example with ESLint v9.39.2, and the behavior of the `no-useless-assignment` rule is the same.

**Comment by barisusakli** [comment]
@mdjermanovic thanks for clarifying, I will adjust our code.

**Comment by nzakas** [comment]
Closing as this isn't a bug (per discussion).

**Comment by jiayc4215** [comment]
> Closing as this isn't a bug (per discussion).

The logic in the example provided is actually fine.The core issue is that the rule triggers a false positive when a variable is defined and assigned in the <script> block, but only referenced within the Vue template.

**Comment by mdjermanovic** [comment]
Indeed the original problem looks different than the example provided later. If anyone can provide a repro repo with a Vue file, please open a new issue. Note that we require a repro repo to be sure what exactly the reported problem is and eventually verify that the problem is fixed. Furthermore, in this case a repro is even more important as there are two third-party parsers involved so the problem may be related to which versions of these parsers were used.
